include/boost/json/impl/value.ipp

```diff
@@ -494,4 +495,54 @@ key_value_pair(
 
 BOOST_JSON_NS_END
 
+//----------------------------------------------------------
+//
+// std::hash specialization
+//
+//----------------------------------------------------------
+
+namespace std {
+std::size_t 
+hash<::boost::json::value>::operator()(
+    ::boost::json::value const& jv) const noexcept 
+{
+  std::size_t seed = static_cast<std::size_t>(jv.kind());
```

Comment:
Should `hash( jv )` == `hash( jv.get_array() )` if `jv.is_array()==true`? With the current code, they will not be equal. I'm not sure if they should be equal or not. @pdimov ?

Reply:
That's a change I requested. Variants typically hash the kind. E.g. the hash of variant<string, string> when it holds the same string as alternative 0 or alternative 1 will differ (which matches what operator== does.)

Reply:
But, as I also said, this as written doesn't match the behavior of value::operator==, where int64 and uint64 are considered equal when they hold the same value.

Reply:
Scratch that, the code further on overrides the seed in the uint64 case. :-) However it assumes that std::hash<int64_t> and std::hash<std::uint64_t> produce the same value, which is not guaranteed.

Reply:
Should `operator==` always return false when comparing int64 to uint64?

Reply:
I don't think there's any pressing need to change the current equality behavior at this time.

Reply:
Question: What is the opinion regarding the equality check with different types (including primitives)? 

For example, these expressions all evaluate to `true`:

```cpp
value(nullptr) == nullptr
value(true) == true
value(false) == false
value(13) == 13
value({1,2,3}) == array({1,2,3})
value({{"a", 1}, {"b",2}}) == object({{"a", 1}, {"b",2}}) 
```
as right hand sides are converted to `value` by `operator==`.

Reply:
That could in theory be quite expensive, can you please open an issue for this in particular?